### ISSUE TITLE
nn: a torch/nn/functional facade for the baseline functional ops

### ISSUE THREAD

**Issue by bkc39** [bkc39/rktorch#112]
The library already has the layering we want — `define-module` is state
management on top of functional routines, and `Conv1d`'s `#:forward` is
literally `(conv1d x weight #:bias bias ...)`. What is missing is a
*name* for the functional layer. It is currently spread across three
places with no unifying module path:

| what | where |
|---|---|
| parameterized ops (`conv1d`, `conv2d`, `max-pool2d`, `avg-pool2d`, `adaptive-avg-pool2d`, `embedding`, `layer-norm`, `masked-fill`, `tril`, `triu`) | `torch/foreign/nn-promoted.rkt` |
| activations (`relu`, `sigmoid`, `gelu`, `tanh`, `softmax`, `log-softmax`) | `torch/foreign/tensor-ops.rkt` via `foreign.rkt` |
| losses (`mse-loss`, `cross-entropy`, `ctc-loss`) | `torch/nn/loss.rkt` |

## Proposal
A `torch/nn/functional` facade that re-exports the baseline functional
operations as one importable unit, mirroring `torch.nn.functional`.

## Design notes, including where I would change the original sketch

**Keep it one file, not a `nn/functional/` directory (for now).** The
sketch had `nn/functional/ops.rkt` re-exporting `foreign/nn-promoted`.
A directory buys nothing yet — `nn-promoted.rkt` is 87 lines and
`loss.rkt` is 49 — and it creates a second importable path
(`torch/nn/functional/ops`) that fragments the surface for no gain.
Start with `torch/nn/functional.rkt`; split into a directory when it
crosses the 500-line target or when there are genuinely separable
families.

**Re-export from the contracted boundary, not the raw module.** This is
load-bearing rather than stylistic: `conv1d` is `define/contract` at its
definition site, while `conv2d` and friends are contracted by
`foreign.rkt`'s `contract-out`. A facade that pulls straight from
`foreign/nn-promoted.rkt` would therefore export some ops contracted and
others bare — two surfaces with different guarantees under one name.
Re-export from `foreign.rkt` (and `nn/loss.rkt`) so every name carries
its contract exactly once.

**The prefix idiom is the point, and it needs documenting.**
`torch/main.rkt` does `(provide (all-from-out "foreign.rkt"))`, so
`relu`, `softmax`, `conv1d` and the rest already come from `torch`.
`(require torch torch/nn/functional)` therefore collides on every shared
name. The intended usage is the direct analogue of
`import torch.nn.functional as F`:

    (require torch (prefix-in F: torch/nn/functional))
    (F:relu (F:conv1d x w #:bias b))

Being prefixable as a unit is most of the facade's value, so the docs
should lead with it.

## Gaps to close as part of this
- **No public functional `linear`.** `Linear`'s forward inlines
  `(add (matmul x (transpose weight 0 1)) bias)`. The most fundamental
  layer is the one you cannot use functionally. Add `linear` to
  `nn-promoted.rkt` and have `Linear` call it.
- **No promoted `dropout`.** It exists only in `torch/generated.rkt`,
  which is documented UNSTABLE; re-exporting that through a stable
  facade would promote it implicitly. Add a promoted wrapper first.

## Open question to settle here, not silently
Three paths would then export the same names: `torch`, `torch/nn`, and
`torch/nn/functional`. Either `torch` keeps them for convenience (and
the facade is purely an ergonomic grouping) or the layering gets
tightened. That decision overlaps #91 and should be made deliberately
rather than inherited.

## Why it matters beyond tidiness
Keeping every module's functional twin complete and named is the cheap
hedge that keeps a functional-first API *buildable* later (params as
explicit data, JAX/Equinox-shaped) without committing to it now. Today
that correspondence is informal, so it erodes silently — `linear` is
already missing.

**Comment by bkc39** [comment]
One coupling not in the original write-up, worth settling before this
is implemented.

**The "re-export from the contracted boundary" advice depends on #96
not having finished yet.** That guidance assumes `foreign.rkt`'s
`contract-out` block is where contracts live. But #96/#101 are actively
migrating the opposite way — contracts to `define/contract-out` at the
definition site, with the facade left re-exporting plain names. `conv1d`
is already on the new side; `conv2d` and the rest are still on the old.

So the facade must not hard-code either assumption. Two consequences:

- Whichever module it pulls from, it should re-export the *binding*, never
  re-wrap with a fresh contract. Wrapping twice gives a name two contract
  layers and blames the facade rather than the caller — the exact
  backwards-blame problem #101 exists to fix.
- The right sequencing is to land this **after** #96's migration settles,
  or to write it so it pulls from wherever the contract currently is and
  accept one follow-up edit when #96 completes. The first is cheaper.

**Cross-reference:** the "three paths export the same names" open
question is really #91's question. Whoever picks up #91 should decide it
there, and this issue should follow that decision rather than making its
own.
